lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/searcher/LocalRepositorySearcher.java

```diff
@@ -69,78 +71,92 @@ public Collection<Artifact> getLocalPluginArtifacts()  {
 	public Collection<Artifact> getLocalArtifactsLastVersion() {
 		CompletableFuture<Collection<Artifact>> loadLocalArtifacts = cache.get(localRepository);
 		if (loadLocalArtifacts == null || loadLocalArtifacts.isCompletedExceptionally()) {
-			// Load local artifacts on background
-			loadLocalArtifacts = CompletableFuture.supplyAsync(() -> {
+			loadLocalArtifacts = getOrCreateLocalArtifactsLastVersion();
+		}
+		if (loadLocalArtifacts.isDone()) {
+			return loadLocalArtifacts.getNow(Collections.emptyList());
+		}
+		// The local artifacts search is not finished, returns an empty list
+		return Collections.emptyList();
+	}
+
+	private synchronized CompletableFuture<Collection<Artifact>> getOrCreateLocalArtifactsLastVersion() {
+		CompletableFuture<Collection<Artifact>> loadLocalArtifacts = cache.get(localRepository);
+		if (loadLocalArtifacts == null || loadLocalArtifacts.isCompletedExceptionally()) {
+			if (MavenLemminxExtension.isUnitTestMode()) {
 				try {
-					return computeLocalArtifacts();
+					loadLocalArtifacts = CompletableFuture.completedFuture(computeLocalArtifacts(() -> {}));
 				} catch (IOException e) {
-					throw new CompletionException(e);
+					LOGGER.log(Level.SEVERE, "Local repo loading error", e);
 				}
-			});
-			
+			} else {
+				// Load local artifacts on background
+				loadLocalArtifacts = CompletableFutures.computeAsync(cancelChecker -> {
+					try {
+						return computeLocalArtifacts(cancelChecker);
+					} catch (IOException e) {
+						throw new CompletionException(e);
+					}
+				});
+			}
+
 			// Update the cache
 			cache.put(localRepository, loadLocalArtifacts);
-			
-			if (MavenLemminxExtension.isUnitTestMode()) {
-				try {
-					loadLocalArtifacts.get();
-				} catch (InterruptedException | ExecutionException e) {
-					Thread.currentThread().interrupt();
-				}
-			}
-			
-			// Once the local artifacts are loaded, we track the local repository to update the cache.
-			loadLocalArtifacts.thenAcceptAsync(artifacts -> {	
+
+			// Once the local artifacts are loaded, we track the local repository to update
+			// the cache.
+			loadLocalArtifacts.thenAcceptAsync(artifacts -> {
 				try {
-				Path localRepoPath = localRepository.toPath();
-				watchService = localRepoPath.getFileSystem().newWatchService();
-				watchKey = localRepoPath.register(watchService, StandardWatchEventKinds.ENTRY_CREATE, StandardWatchEventKinds.ENTRY_DELETE);
-				
-				WatchKey key;
-				
+					Path localRepoPath = localRepository.toPath();
+					watchService = localRepoPath.getFileSystem().newWatchService();
+					watchKey = localRepoPath.register(watchService, StandardWatchEventKinds.ENTRY_CREATE,
+							StandardWatchEventKinds.ENTRY_DELETE);
+
+					WatchKey key;
+
 					while ((key = (watchService != null ? watchService.take() : null)) != null) {
 						if (watchKey.equals(key)) {
 							cache.remove(localRepository);

```

Comment:
@angelozerr You haven't added any calls to start loading an updated local repo data here. This means a user will most-likely  get an empty artifact list on  next request and local repo loading will be started in background.
Why not to start loading a new local repo right here, so at the moment of new user request it will be propably already loaded and cached?

Reply:
To be honnest with you, I have just keep the old code which was done in a Thread.

Reply:
We could improve that in the future.

Reply:
OK. Agreed